vsphere/resource_vsphere_virtual_machine_test.go

```diff
@@ -8088,3 +8129,188 @@ resource "vsphere_virtual_machine" "vm" {
 		os.Getenv("VSPHERE_TEMPLATE"),
 	)
 }
+
+func testAccResourceVSphereVirtualMachineConfigBasicEmptyCluster() string {
+	return fmt.Sprintf(`
+variable "datacenter" {
+  default = "%s"
+}
+
+variable "resource_pool" {
+  default = "%s/Resources"
+}
+
+variable "network_label" {
+  default = "%s"
+}
+
+variable "datastore" {
+  default = "%s"
+}
+
+data "vsphere_datacenter" "dc" {
+  name = "${var.datacenter}"
+}
+
+data "vsphere_datastore" "datastore" {
+  name          = "${var.datastore}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_resource_pool" "pool" {
+  name          = "${var.resource_pool}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+data "vsphere_network" "network" {
+  name          = "${var.network_label}"
+  datacenter_id = "${data.vsphere_datacenter.dc.id}"
+}
+
+resource "vsphere_virtual_machine" "vm" {
+  name             = "terraform-test"
+  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"
+  datastore_id     = "${data.vsphere_datastore.datastore.id}"
+
+  num_cpus = 2
+  memory   = 2048
+  guest_id = "other3xLinux64Guest"
+
+  network_interface {
+    network_id = "${data.vsphere_network.network.id}"
+  }
+
+  disk {
+    label = "disk0"
+    size  = 20
+  }
+}
+`,
+		os.Getenv("VSPHERE_DATACENTER"),
+		os.Getenv("VSPHERE_EMPTY_CLUSTER"),
+		os.Getenv("VSPHERE_NETWORK_LABEL_PXE"),
+		os.Getenv("VSPHERE_DATASTORE"),
```

Comment:
nit: these functions feel like they'd be more reusable if the environment variables are passed in as arguments, but I don't have the context to know if that's a valid design consideration or not.

Reply:
Where applicable, configuration generators do have parameterization, but again, only when it's necessary in a test (ie: a multi-step test where the configuration is the same but a variable is changing). Otherwise, these configurations should be considered static and doubling up the parameterization would only serve to make things more redundant.

Some future work I'm considering doing is possibly adding to the test suite so that we can load external configuration (ie: out of a `test-fixtures` directory), and allowing the use of actual variables. If that indeed comes to fruition, this is probably when I would come back to this and refactor things so that the parameterization is in the individual test cases themselves.

Reply:
on that note - tbh there's an argument for making these values a Struct that's parsed out once, rather than in each test function (this is the case in all providers)

Reply:
I would be fine with this if there was a way to wrap it in a way that retains the same "static generation" look and feel. I actually am not a fan of having variables declared in _every_ test function.

I think that something along the lines of this would work even:

```
Config: resource.GenerateConfig(testAccResourceVSphereConfigHere),
```

Where `resource.GenerateConfig` is a generator that takes a `text/template` with a lightweight, Packer-ish way of accessing environment variables.

```go
testAccSomeConfig = `
variable "foo" {
  default = "{{env TEST_PROVDER_ENV_VAR}}"
}
`
```

In any case, such things are kind of out of scope here :stuck_out_tongue: